src/zarr/core/group.py

```diff
@@ -2627,3 +2600,191 @@ def array(
                 )
             )
         )
+
+
+async def _getitem_semaphore(
```

Comment:
Is there a reason we need to passing a semaphore around?


FWIW we have this utility function in Arraylake that gets decently wide use.

```python
async def async_gather(*coros, concurrency: Optional[int] = None, return_exceptions: bool = False) -> list[Any]:
    """Execute a gather while limiting the number of concurrent tasks.

    Args:
        coros: coroutines
            list of coroutines to execute
        concurrency: int
            concurrency limit
            if None, defaults to config_obj.get('async.concurrency', 4)
            if <= 0, no concurrency limit

    """
    if concurrency is None:
        concurrency = int(config_obj.get("async.concurrency", 4))

    if concurrency > 0:
        # if concurrency > 0, we use a semaphore to limit the number of concurrent coroutines
        semaphore = asyncio.Semaphore(concurrency)

        async def sem_coro(coro):
            async with semaphore:
                return await coro

        results = await asyncio.gather(*(sem_coro(c) for c in coros), return_exceptions=return_exceptions)
    else:
        results = await asyncio.gather(*coros, return_exceptions=return_exceptions)

    return results
```

Reply:
I'm up for implementing an `as_completed_with_sempahore` function, as I'm sure we will need that in more places. But i'm not sure if that should be a blocker for this pr